src/main/java/uk/gov/hmcts/reform/iacaseapi/domain/handlers/presubmit/AppealReferenceNumberHandler.java

```diff
@@ -0,0 +1,69 @@
+package uk.gov.hmcts.reform.iacaseapi.domain.handlers.presubmit;
+
+import static java.util.Arrays.asList;
+import static java.util.Objects.requireNonNull;
+
+import java.util.Optional;
+import org.springframework.stereotype.Service;
+import uk.gov.hmcts.reform.iacaseapi.domain.entities.AsylumCase;
+import uk.gov.hmcts.reform.iacaseapi.domain.entities.ccd.Event;
+import uk.gov.hmcts.reform.iacaseapi.domain.entities.ccd.callback.Callback;
+import uk.gov.hmcts.reform.iacaseapi.domain.entities.ccd.callback.PreSubmitCallbackResponse;
+import uk.gov.hmcts.reform.iacaseapi.domain.entities.ccd.callback.PreSubmitCallbackStage;
+import uk.gov.hmcts.reform.iacaseapi.domain.handlers.PreSubmitCallbackHandler;
+import uk.gov.hmcts.reform.iacaseapi.domain.service.AppealReferenceNumberGenerator;
+
+@Service
+public class AppealReferenceNumberHandler implements PreSubmitCallbackHandler<AsylumCase> {
+
+    private final AppealReferenceNumberGenerator appealReferenceNumberGenerator;
+
+    public AppealReferenceNumberHandler(
+            AppealReferenceNumberGenerator appealReferenceNumberGenerator) {
+
+        this.appealReferenceNumberGenerator = appealReferenceNumberGenerator;
+    }
+
+    @Override
+    public boolean canHandle(PreSubmitCallbackStage callbackStage, Callback<AsylumCase> callback) {
+
+        requireNonNull(callbackStage, "callbackStage must not be null");
+        requireNonNull(callback, "callback must not be null");
+
+        return callbackStage == PreSubmitCallbackStage.ABOUT_TO_SUBMIT
+                && callback.getEvent() == Event.SUBMIT_APPEAL;
+    }
+
+    @Override
+    public PreSubmitCallbackResponse<AsylumCase> handle(
+            PreSubmitCallbackStage callbackStage,
+            Callback<AsylumCase> callback) {
+
+        if (!canHandle(callbackStage, callback)) {
+            throw new IllegalStateException("Cannot handle callback");
+        }
+
+        AsylumCase caseData = callback
+                .getCaseDetails()
+                .getCaseData();
+
+        PreSubmitCallbackResponse<AsylumCase> callbackResponse =
+                new PreSubmitCallbackResponse<>(caseData);
+
+        if (!caseData.getAppealReferenceNumber().isPresent()) {
+
+            Optional<String> appealReferenceNumber =
+                appealReferenceNumberGenerator.getNextAppealReferenceNumberFor(caseData.getAppealType().get());
+
+            if (!appealReferenceNumber.isPresent()) {
+                callbackResponse.addErrors(asList("Sorry, there was a problem submitting your appeal case"));
+            } else {
+                caseData.setAppealReferenceNumber(appealReferenceNumber.get());
+            }
+        } else {
+            callbackResponse.addErrors(asList("Sorry, there was a problem submitting your appeal case"));
```

Comment:
We can maybe do this tomorrow together, but this needs to be an exception.
As it stands, the dispatcher would carry on calling other handlers, which may try to generate documents or send email notifications etc.